### ISSUE TITLE
Long loading times on GeneratePdf via Azure App Service

### ISSUE THREAD

**Issue by FritschP** [QuestPDF/QuestPDF#911]
**Describe the bug**
We currently have a problem on our Azure App Service running QuestPDF. Testing locally our report works great, the generation process is fast and reliable. However, we deployed our code to production (which is running on an Azure App Service) and there the first PDF generation process takes a really long time (sometimes 15 to 30 minutes). After the first PDF is generated sucessfully, each following PDF is generated in seconds. This works as long as the Azure App Service is up and running - after a restart this behaviour starts again.
Our workaround for this issue is now a "warmup" which starts a PDF generation process after starting the server. 

**To Reproduce**
We used a extremly basic report with the following code and the behaviour is still the same. This can be reproduced on our production system which is running in Azure.
```
        public void Compose(IDocumentContainer container)
        {
            container
                .Page(page =>
                {
                    page.Margin(50);
                });
        }
```

**Expected behavior**
Fast PDF generation

**Environment**
QuestPDF Version: 2024.3.1
Environment: Azure App Service running on a Premium V3 (P1V3) Hardware, Dotnet - v6.0 (x64), Hosting as OutOfProcess

**Comment by MarcinZiabek** [comment]
Before we start the investigation, could you please update the library to the latest version and check if it solves the issue?

**Comment by FritschP** [comment]
Thanks for coming back to us. Here are our answers to your question:
1. Yes this behaviour is specific to the PDF generation process. Other endpoints are working as intended.
2. Not that we know of.
3. The PDF document generation happens indeed in a service which is used via Dependency Injection, however no sort of dependencies are costly to create as this service receives a model (in this dummy case the model is null) and generates the corresponding PDF via QuestPDF without any more dependencies.
4. No, everything is working as intended on localhost

**Comment by MarcinZiabek** [comment]
Thank you for your answers.

Where did you put this line of code? Is it in the PDF generation service or in the startup file?

```csharp
Settings.License = LicenseType.Community;
```

Could you please remove any PDF generation logic, and put this line only in the generation service?

Explanation: the location does not matter. However, this line should force the managed code to load QuestPDF native dependencies. I wonder if there is any antivirus/firewall/monitoring software on your App Service that could require a lot of time to scan this dll file. I find it surprising how consistent is the observed behavior

**Comment by ursaciuc-adrian** [comment]
I am having the same issue, although not to the same extent. The first generation takes 10 seconds then the rest would take 0.5 seconds.

**Comment by FritschP** [comment]
@MarcinZiabek sorry for the late response, we could only test this on a specific Azure Environment and there was no time to test it extensivly. However, now we gathered more intel. Adding the license line works without any trouble, this can be done via the startup file or the generation service, the result is the same as it works without any problems.

Generating a pdf with only the following code also works without any delay or problems.
```
  container.Page(page =>
    {
        page.Size(PageSizes.A4);
        page.Margin(2, Unit.Centimetre);
        page.PageColor(Colors.White);
        page.DefaultTextStyle(x => x.FontSize(20));
    }
```

However, adding a content line will create the problem mentioned above, resulting in a extremly long loading time followed by everthing working normally. We tested this via adding the following code to the document generation process, taken from the "Quick start" documentation.
 ```
page.Header()
            .Text("Hello PDF!")
            .SemiBold().FontSize(36).FontColor(Colors.Blue.Medium);
```

Could you please guide us how to troubleshoot it further?

**Comment by chaimtevel** [comment]
Hi,

Thanks for building and maintaining this excellent library!

We've been seeing very similar issues.

We're running linux docker containers on azure container apps. The first run takes at least two minutes (it times out after that) but subsequent runs work quite nicely.

We do have lots of files in the root of our app (`AppDomain.CurrentDomain.BaseDirectory` - thats the location that `Helpers.Helpers.ApplicationFilesPath` falls back on) as we use [Azure file mounting](https://learn.microsoft.com/en-us/azure/container-apps/storage-mounts?tabs=smb&pivots=azure-cli#azure-files) to manage files and thats mounted at the root of our app.

We're unable to delete these most of these files as these files are used by our application but I did delete many of them and now the initial run takes about 50 seconds.

Is there a setting that can control where QuestPDF looks for these fonts? Or another setting that can minimize the work that needs to happen on startup?

Thanks!

**Comment by FritschP** [comment]
@MarcinZiabek Thanks for the feedback, it is indeed this part of the code! Particular this line:
```
var fontFilePaths = supportedFontExtensions
                .SelectMany(extension => Directory.GetFiles(Helpers.Helpers.ApplicationFilesPath, extension, SearchOption.AllDirectories));
```
The reason for this was, that we had a local cache built on the filesystem of our production environment which had already over 15.000 folders created. We recently moved to Azure Blob Storage, as we noticed that many requests regarding the file-system are particular slow in Azure App Services, most likely to the nature of storage not being directly on the machine, but accessed through the network instead. This cache however was not cleaned up and persisted on the filesystem. 

We tested it, before the cleanup this line of code resulted in a request timeout (>3minuts 50 seconds) and now it completes in under 1 second.  After restarting the Service, the pdf generation also took the expected amount of time - so this fixed it for us!

I think it would be nice if the path for searching for font files would be configurable, or to document this behaviour into a "Attention" section. However, we greatly appreciated the help and the fast responses!

**Comment by MarcinZiabek** [comment]
Before introducing any configuration to this feature (toggle to enable/disable it, or configure the search path), I would like to first investigate if the code could be made faster. I have put this item on my to-do list 😄

**Comment by chaimtevel** [comment]
Ok great!

If you know of any other workarounds that we can use in the interim (other than the location where we mount files which can get sticky with permissions) then please let us know.

**Comment by MarcinZiabek** [comment]
Good question! As far as I see, there is no c# API that could be more performant. Do you know of any better solutions? Maybe listing all files and then filtering in c#? But listing 50k files will undoubtedly be slow... (yet to be benchmarked) 😆 

The last resort option is to create a configurable setting for that. However, having so many files in the application directory is likely not normal and indicates a larger problem with the design.

**Comment by chaimtevel** [comment]
I do think your suggestion as to filtering in code directly may be worth a shot. The code [here](https://github.com/QuestPDF/QuestPDF/blob/87e373f682f5d8245b0936a1ccde9402cf119791/Source/QuestPDF/Drawing/FontManager.cs#L73) is scanning the file hierarchy once per file extension which in this case is four times. I'm not sure, but I'd think its worth exploring scanning the hierarchy just once and find the relevant files in code.

From a design perspective, from what I understand its not uncommon for app's that generate files to store them under the web site root. In our case its actually just a cloud share so we have the flexibility of changing the mount to another location but for others changing the location may not be so simple.

As an aside, I think giving developers some form of control of these sort of things is likely going to help them. Having a default behavior may likely be good for the majority of use-cases but allowing overrides allows for more flexibility.

For now we'll just work on changing where we mount the shared drive.

Thanks for looking into this!

**Comment by rolighed** [comment]
Any news in regards to this? In our solution we manually register the necessary specific fonts (FontManager.RegisterFont) which is located in a subfolder, so the possibility to disable the automatic scanning would be _greatly_ appreciated!

**Comment by MarcinZiabek** [comment]
Yes 😄 The `2024.7.3` release introduces a new `QuestPDF.Settings.FontDiscoveryPaths` setting. You can specify paths the library scans for font files and registers them. Moreover, if the total number of files in the specified path exceeds a certain threshold, the library throws an exception pointing to this solution. Could you please test it?

**Comment by chaimtevel** [comment]
Hi,

Though I have moved most of our directories outside of the app's base directory so this issue was no longer relevant to us, I was in this code again so I spent a few minutes testing

I can confirm that doing the below does fix the issue.

```cs
QuestPDF.Settings.FontDiscoveryPaths.Clear();
QuestPDF.Settings.FontDiscoveryPaths.Add(@"path/to/font/directory");
```

Thanks for adding this quickly!

**Comment by MarcinZiabek** [comment]
Since there has been no recent activity, I’m closing this issue. If you experience a similar problem or have a related request, feel free to open a new GitHub issue. I appreciate your feedback! ❤️
